### PR TITLE
Wrap R to enable 'backcasting' validation

### PR DESCRIPTION
**Pull request by henrygarner**
Call out to R so that helper models can be programmatically trained on
different subsets of data. This allows us to validate our model with
backcasting - projecting historic periods - so that we can compare our
model projections to the historic actuals. We also train a simple
linear regression model on the same subsets of data for an additional comparison.

### CODE REVIEW

File: src/cic/core.clj [hunk 1]
```diff
@@ -1,184 +1 @@
...
-                    (into {}))]
-    {:ages ages :params params}))
-
-(defn load-costs-csv
-  [filename]
-  (->> (load-csv filename)
-       (map #(-> %
-                 (update :placement keyword)
-                 (update :cost parse-double)))))
+(ns cic.core)
```
Comment (seb231) [inline src/cic/core.clj]: Delete file? Unless it is re-implemented in a later PR?

File: src/cic/repl.clj [hunk 2]
```diff
@@ -1,86 +1,89 @@
...
-  (let [periods (core/episodes->periods episodes)]
-    (projection/projection periods
-                           project-from project-to
-                           joiners-model duration-model
-                           seed
-                           10)))
+(defn load-model-inputs
+  "A useful REPL function to load the data files and convert them to  model inputs"
+  []
+  (hash-map :periods (-> (read/episodes "data/episodes.csv")
```
Comment (seb231) [inline src/cic/repl.clj]: So if this is reading in episodes.csv perhaps we should translate the Python code that currently preps this file from the SSDA903 data, @otfrom ?

File: src/cic/io/read.clj [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                     (->> in-file
+                          data-csv/read-csv
+                          (remove blank-row?)
+                          (vec)))
+        parsed-data (rest parsed-csv)]
+    (->> (map (comp (juxt first (comp vec rest)) (partial map parse-int)) parsed-data)
+         (into {}))))
+
+(defn duration-csvs
+  [lower median upper]
```
Comment (seb231) [inline src/cic/io/read.clj]: pop in a docstring for what the fn actually expects

File: src/cic/io/read.clj [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                 median (get median age)
+                 upper (get upper age)]
+             (vector age
+                     (mapv (fn [l m u]
+                             (vector l m u))
+                           lower median upper))))
+         (into {}))))
+
+(defn joiner-csvs
+  [mvn params]
```
Comment (seb231) [inline src/cic/io/read.clj]: and here

File: src/cic/repl.clj [hunk 5]
```diff
@@ -1,86 +1,89 @@
...
   (-> (assoc summary :date date)
       (cs/rename-keys {:total :actual})))
 
-(defn assoc-costs
-  [placement-costs placement-counts]
-  (let [cost (reduce (fn [total {:keys [placement cost]}]
-                       (+ total (* cost (get placement-counts placement 0))))
-                     0 placement-costs)]
-    (assoc placement-counts :cost cost)))
+(defn write-projection!
```
Comment (seb231) [inline src/cic/repl.clj]: I'm not sure this is quite the right name, as "writing" is not the only thing it does

File: src/joiners.R [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+args = commandArgs(trailingOnly=TRUE)
+input <- args[1]
+model.out <- args[2]
+rates.out <- args[3]
+seed.long <- args[4]
+set.seed(seed.long)
+
+# input <- '/var/folders/tw/tzkjfqd11md5hdjcyjxgmf740000gn/T/file726180705789120480.csv'
+# model.out <- '/var/folders/tw/tzkjfqd11md5hdjcyjxgmf740000gn/T/file8865703702117111587.csv'
+# rates.out <- '/var/folders/tw/tzkjfqd11md5hdjcyjxgmf740000gn/T/file2591792755455537745.csv'
```
Comment (seb231) [inline src/joiners.R]: ??

File: src/joiners.R [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+df <- read.csv(input, header = TRUE, stringsAsFactors = FALSE, na.strings ='')
+df$beginning <- as.Date(parse_date_time(df$beginning, 'ymd HMS'))
+df$admission_age <- as.factor(as.character(df$admission_age))
+
+diffs <- df %>% arrange(admission_age, beginning) %>% group_by(admission_age) %>%
+  mutate(diff = interval(lag(beginning), beginning) / days(1), n = n()) %>%
+  ungroup %>% filter(!is.na(diff) & n >= 3) %>% # We need at least 3 data points for each age to generate 2 diffs
+  dplyr::select(admission_age, diff, beginning) %>% mutate(diff = diff + 0.01) %>%
+  mutate(admission_age = droplevels(admission_age)) %>%
+  as.data.frame
```
Comment (seb231) [inline src/joiners.R]: sorry for being pedantic, could you follow each pipe (`%>%`) with a newline for ease of reading please

File: src/cic/repl.clj (hunk 2 above)
Comment (otfrom) [reply src/cic/repl.clj]: @seb231 yes we should. I'd like to reduce the number of tools we need to use to get to where we want to be.

File: src/cic/repl.clj [hunk 8]
```diff
@@ -1,86 +1,89 @@
...
-        projection (->> (concat summary-seq
-                                (project episodes project-from project-to
-                                         joiners-model duration-model seed))
-                        (map (partial assoc-costs placement-costs)))]
-    (write-projection-tsv output-file projection)))
+(defn write-validation!
+  "Outputs model projection and linear regression projection together with actuals for comparison."
+  [out-file n-runs seed]
+  (let [{:keys [periods placement-costs duration-model]} (load-model-inputs)]
+    (->> (for [as-at (time/month-seq (time/make-date 2010 1 1)
```
Comment (otfrom) [inline src/cic/repl.clj]: why a for rather than a fn with a let inside?

File: src/cic/repl.clj (hunk 2 above)
Comment (seb231) [reply src/cic/repl.clj]: Discussed elsewhere. Need to translate R and Python code to single independent Clojure tool for prepping SSDA903

**Comment by seb231** [comment]
Tests to not need to pass, as they will be fixed in future branch

**Pull request merged**